bin/update_version_number_in_files.py

```diff
@@ -0,0 +1,212 @@
+        r"^set\(_bout_previous_version \"v(\d+\.\d+\.\d+)\"\)",
+        new_version_number,
+    )
+    update_version_number_in_file(
+        "CMakeLists.txt",
+        r"^set\(_bout_next_version \"(\d+\.\d+\.\d+)\"\)",
+        bout_next_version_number,
+    )
+    update_version_number_in_file(
+        "tools/pylib/_boutpp_build/backend.py",
+        r"_bout_previous_version = \"v(\d+\.\d+\.\d+)\"",
+        new_version_number,
+    )
+    update_version_number_in_file(
+        "tools/pylib/_boutpp_build/backend.py",
+        r"_bout_next_version = \"v(\d+\.\d+\.\d+)\"",
+        bout_next_version_number,
+    )
+
+
+@dataclass
+class VersionNumber:
+    major_version: int
+    minor_version: int
+    patch_version: int
+
+    def __str__(self):
+        return "%d.%d.%d" % (self.major_version, self.minor_version, self.patch_version)
+
+
+@dataclass
+class ShortVersionNumber:
+    major_version: int
+    minor_version: int
+
+    def __str__(self):
+        return "%d.%d" % (self.major_version, self.minor_version)
+
+
+def apply_fixes(pattern, new_version_number, source):
+    """Apply the various fixes for each factory to source. Returns
+    modified source
+
+    Parameters
+    ----------
+    pattern
+        Regex pattern to apply for replacement
+    new_version_number
+        New version number to use in replacement
+    source
+        Text to update
+    """
+
+    def get_replacement(match):
+        return match[0].replace(match[1], str(new_version_number))
+
+    modified = re.sub(pattern, get_replacement, source, flags=re.MULTILINE)
+
+    return modified
+
+
+def yes_or_no(question):
+    """Convert user input from yes/no variations to True/False"""
+    while True:
+        reply = input(question + " [y/N] ").lower().strip()
+        if not reply or reply[0] == "n":
+            return False
+        if reply[0] == "y":
+            return True
+
+
+def create_patch(filename, original, modified):
+    """Create a unified diff between original and modified"""
+
+    patch = "\n".join(
+        difflib.unified_diff(
+            original.splitlines(),
+            modified.splitlines(),
+            fromfile=filename,
+            tofile=filename,
+            lineterm="",
+        )
+    )
+
+    return patch
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description=textwrap.dedent(
+            """\
+            Update the software version number to the specified version, 
+            to be given in the form major.minor.patch, 
+            e.g. 5.10.3
+            
+            Where the 3rd ('patch') part of the version is omitted, 
+            only the 'major' and 'minor' parts will be used, 
+            e.g. 5.10.3 -> 5.10 
+            
+            For the 'bout-next' version number, 
+            the 'minor' version number of the provided version will be incremented by 1, 
+            e.g. 5.10.3 -> 5.11.3
```

Comment:
But that is not what we want?

For master we want to bump the patch level
for next we want to bump the minor or major.

And always, we want to reset all less significant numbers to zero.

I think we might not need to provide the version we want to bump to, because that has already been done, but rather say what the next version bump is going to be.

So when we want to release say 5.1.0 we say we want 5.1.1 for the master branch and 5.2.0 for the next branch, as 5.1.0 is already stored as the future version ( I think in CMakeLists.txt and backend.py)

Reply:
All good points! Might need some more involved logic here